common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -593,6 +607,9 @@ public void onData(String streamId, ByteBuffer buf) {
 
         @Override
         public void onComplete(String streamId) {
+          if (isTooLate) {
+            pushMergeMetrics.tooLateResponses.mark();
+          }
```

Comment:
li-2.3.0 has the corresponding implementation as below:
```
public void onComplete(String streamId) {
  if (isTooLate) {
    pushMergeMetrics.tooLateResponses.mark();
    // Throw an exception here so the block data is drained from channel and server
    // responds RpcFailure to the client.
    throw new RuntimeException(String.format("Block %s %s", msg.blockId,
      BlockPushException.TOO_LATE_MESSAGE_SUFFIX));
  }
  // For duplicate block that is received before the shuffle merge finalizes, the
  // server should respond success to the client.
}
```
I think the reason we are doing this only as part of `onComplete` is that we only mark tooLateResponse when all data from the stream has been received, and ignore those for the callbacks of `onData` and `onFailure`, correct? @otterc

Reply:
That looks like an impl detail in li-2.3.0 branch - any reason not to update this sooner ?

Reply:
This is explained in the comments above that why we don't want throw exception and close the channel but rather respond with stream callback and only when the complete block is received, respond with tooLate

Reply:
Yes, but is there a reason to defer updating the metric to later ? It is clear at time of callback construction itself that it is late.

Reply:
Any comments on this @otterc, @zhouyejoe ?
Essentially, we are not marking late block push in case of `onFailure` - any reason for it ?

IMO we should do it for both - and if yes, we can move it out of the callback entirely and do it immediately before the callback construction.
Pls let me know if I am missing something ! Thx

Reply:
Actually in our 2.3 version, we were incrementing this metric whenever, the server was responding to the client that the block is tooLate. In `onFailure` we don't respond to the client with `TooLate`. However, these metrics are the server's view so I think you are right. We should increment outside the callback.

Reply:
Can you make this change @rmcyang  ? Thx

Reply:
Thanks for the comments. Fixed.